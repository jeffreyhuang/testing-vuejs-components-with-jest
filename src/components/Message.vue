<template>
  <li
    style="margin-top: 10px"
    class="message"
    data-test="test-message"
  >
    <span>
      {{ message }}
    </span>
  </li>
</template>

<script>
export default {
  props: ['message']
}
</script>
