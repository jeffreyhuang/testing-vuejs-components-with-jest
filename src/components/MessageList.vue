<template>
  <ul>
    <Message
      v-for="(message, index) in messages"
      :key="index"
      :message="message"
    />
  </ul>
</template>

<script>
import Message from './Message'

export default {
  name: 'MessageList',
  components: {
    Message
  },
  props: ['messages']
}
</script>
