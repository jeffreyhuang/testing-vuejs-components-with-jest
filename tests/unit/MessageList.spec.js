// working code:
// https://github.com/alexjoverm/vue-testing-series/blob/Test-Styles-and-Structure-in-Vue-js-and-Jest/test/MessageList.test.js

import { mount } from '@vue/test-utils'
import MessageList from '@/components/MessageList'
import Message from '@/components/Message'

describe('MessageList.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(MessageList, {
      propsData: {
        messages: ['Cat']
      }
    })
  })

  it('has received ["Cat"] as the message property', () => {
    expect(wrapper.vm.messages).toEqual(['Cat'])
  })

  it('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('is a MessageList component', () => {
    expect(wrapper.is(MessageList)).toBe(true)

    // CSS selector
    expect(wrapper.is('ul')).toBe(true)
  })

  it('contains a Message component', () => {
    expect(wrapper.contains(Message)).toBe(true)

    // CSS selector
    expect(wrapper.contains('.message')).toBe(true)
  })

  // Vue Instance
  it('Both MessageList and Message are vue instance', () => {
    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.find(Message).isVueInstance()).toBe(true)
  })

  it('Message has a "message" property equals to "Cat', () => {
    expect(wrapper.find(Message).props('message')).toBe('Cat')
  })

  // Structure
  it('Message element exists', () => {
    expect(wrapper.find('.message').exists()).toBe(true)
  })

  it('Message is not empty', () => {
    expect(wrapper.find(Message).isEmpty()).toBe(false)
  })

  it("Message has a class attribute set to 'message'", () => {
    expect(wrapper.find(Message).attributes('class')).toBe('message')

    // data attributes
    expect(wrapper.find(Message).attributes('data-test')).toBe('test-message')
  })

  // Style
  it('Message component has the .message class', () => {
    expect(wrapper.find(Message).classes()).toContain('message')
  })

  it('Message component has style margin-top 10px', () => {
    expect(wrapper.find(Message).attributes('style')).toBe('margin-top: 10px;')
  })
})
