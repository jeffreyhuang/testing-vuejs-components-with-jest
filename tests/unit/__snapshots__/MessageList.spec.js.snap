// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList.vue has the expected html structure 1`] = `
<ul>
  <li
    class="message"
    data-test="test-message"
    style="margin-top: 10px;"
  >
    <span>
      
    Cat
  
    </span>
  </li>
</ul>
`;
